src/core/buffers.rs

```diff
@@ -195,34 +195,119 @@ impl<B> AsMut<B> for DynamicStorageBuffer<B> {
 }
 
 impl<B: BufferMut> DynamicStorageBuffer<B> {
+    /// Layouts and writes an entire bound value into the buffer. The value is written at the
+    /// next available offset after the buffer's current offset, which is aligned to the required
+    /// dynamic binding alignment (defaults to 256).
+    ///
+    /// Use this to write the entire struct you will be binding as a dynamic-offset storage buffer.
+    ///
+    /// Returns the offset at which the value was written.
     pub fn write<T>(&mut self, value: &T) -> Result<u64>
     where
         T: ShaderType + WriteInto,
     {
+        self.write_dynamic_struct_break();
         let offset = self.offset;
 
         let mut writer = Writer::new(value, &mut self.inner, offset)?;
         value.write_into(&mut writer);
 
-        self.offset += self.alignment.round_up(value.size().get()) as usize;
+        self.offset += value.size().get() as usize;
+
+        Ok(offset as u64)
+    }
+
+    /// Layouts and writes a single member into the buffer. The value is written at the
+    /// next available offset after the buffer's current offset, which is aligned the
+    /// alignment of `T`.
+    ///
+    /// The use case is constructing a struct member by member in case the layout isn't
+    /// known at compile time. Combine this with [`Self::write_dynamic_struct_break`] to "end"
+    /// the struct being written and start the next one.
+    ///
+    /// Returns the offset at which the value was written.
+    pub fn write_single_member<T>(&mut self, value: &T) -> Result<u64>
+    where
+        T: ShaderType + WriteInto,
+    {
+        self.offset = T::METADATA.alignment().round_up(self.offset as u64) as usize;
+        let offset = self.offset;
+
+        let mut writer = Writer::new(value, &mut self.inner, offset)?;
+        value.write_into(&mut writer);
+
+        self.offset += value.size().get() as usize;
 
         Ok(offset as u64)
     }
+
+    /// Writes a "struct break" into the buffer. This takes the buffer offset and aligns it
+    /// to the required dynamic binding alignment (defaults to 256).
+    ///
+    /// The use case is constructing a struct member by member in case the layout isn't
+    /// known at compile time. Combine this with [`Self::write_single_member`] to add
+    /// each individual member of the struct being written.
+    ///
+    /// Returns the offset which was rounded up to.
+    pub fn write_dynamic_struct_break(&mut self) -> u64 {
+        self.offset = self.alignment.round_up(self.offset as u64) as usize;
+        self.offset as u64
```

Comment:
> This might not be enough, some structs needs padding between the last field and the struct end; we probably have to keep track of the alignment of the current struct and pad here explicitly.

If self.alignment > needed_alignment (which we have asserted to be true during construction of a dynamic buffer as needed_alignment can't be more than 32), then we should always be good here.

Reply:
Right, but the issue is that if the buffer itself is smaller than the size of the struct, we don't currently enlarge it to be the size of the struct (we only increment the offset).

Reply:
Should have added code to deal with this.

Reply:
Looks good! What do you think of erroring on empty structs?

Reply:
I'n not sure it's necissarily the responsibility of this to detect that?